File: 0416-partition-equal-subset-sum/0416-partition-equal-subset-sum.java
class Solution {
    public boolean canPartition(int[] nums) {
        int totalSum = 0;
        
        for (int num : nums) {
            totalSum += num;
        }

        if (totalSum % 2 != 0) {
            return false;
        }

        int target = totalSum / 2;
        int n = nums.length;

        boolean[] dp = new boolean[target + 1];
        dp[0] = true; 

        for (int num : nums) {
            for (int t = target; t >= num; t--) {
                dp[t] = dp[t] || dp[t - num];
            }
        }

        return dp[target];
    }
}

File: 0416-partition-equal-subset-sum/0416-partition-equal-subset-sum.py
class Solution:
    def canPartition(self, nums: List[int]) -> bool:
        total_sum = sum(nums)

        # If the total sum is odd, it's not possible to partition it into two equal subsets
        if total_sum % 2 != 0:
            return False

        target = total_sum // 2
        n = len(nums)

        # DP array to track possible sums
        dp = [False] * (target + 1)
        dp[0] = True  # Zero sum is always possible

        for num in nums:
            # Iterate backward to avoid overwriting the dp array
            for t in range(target, num - 1, -1):
                dp[t] = dp[t] or dp[t - num]

        return dp[target]
